plugins/obs-ffmpeg/obs-ffmpeg-vaapi.c

```diff
@@ -160,19 +160,67 @@ static bool vaapi_init_codec(struct vaapi_encoder *enc, const char *path)
 	return true;
 }
 
+/* "Allowed" options per Rate Control
+ * See FFMPEG libavcodec/vaapi_encode.c (vaapi_encode_rc_modes)
+ */
+typedef struct {
+	const char *name;
+	bool qp;
+	bool bitrate;
+	bool maxrate;
+} rc_mode_t;
+
+static const rc_mode_t *get_rc_mode(const char *name)
+{
+	/* Set "allowed" options per Rate Control */
+	static const rc_mode_t RC_MODES[] = {
+		{.name = "CBR", .qp = false, .bitrate = true, .maxrate = false},
+		{.name = "CQP", .qp = true, .bitrate = false, .maxrate = false},
+		{.name = "VBR", .qp = false, .bitrate = true, .maxrate = true},
+		NULL};
+
+	const rc_mode_t *rc_mode = RC_MODES;
+
+	while (!!rc_mode && strcmp(rc_mode->name, name) != 0)
+		rc_mode++;
+
+	return rc_mode ? rc_mode : RC_MODES;
+}
+
 static bool vaapi_update(void *data, obs_data_t *settings)
 {
 	struct vaapi_encoder *enc = data;
 
 	const char *device = obs_data_get_string(settings, "vaapi_device");
 
+	const char *rate_control =
+		obs_data_get_string(settings, "rate_control");
+	const rc_mode_t *rc_mode = get_rc_mode(rate_control);
+	bool cbr = strcmp(rc_mode->name, "CBR") == 0;
+
 	int profile = (int)obs_data_get_int(settings, "profile");
 	int bf = (int)obs_data_get_int(settings, "bf");
+	int qp = rc_mode->qp ? (int)obs_data_get_int(settings, "qp") : 0;
+
+	av_opt_set_int(enc->context->priv_data, "qp", qp, 0);
 
 	int level = (int)obs_data_get_int(settings, "level");
-	int bitrate = (int)obs_data_get_int(settings, "bitrate");
+	int bitrate = rc_mode->bitrate
+			      ? (int)obs_data_get_int(settings, "bitrate")
+			      : 0;
+	int maxrate = rc_mode->maxrate
+			      ? (int)obs_data_get_int(settings, "maxrate")
+			      : 0;
 	int keyint_sec = (int)obs_data_get_int(settings, "keyint_sec");
 
+	/* For Rate Control which allows maxrate, insure maxrate > bitrate
+	 * otherwise, FFMPEG will give an error.
+	 * For CBR, maxrate = bitrate
+	 */
+	if (cbr || (rc_mode->maxrate && maxrate && maxrate < bitrate)) {
```

Comment:
For VBR the maxrate should be 0 or a number greater than the bitrate. Setting maxrate=bitrate here in the VBR case (where maxrate < bitrate) causes ffmpeg to choose CBR rate control.

Reply:
For VBR, what should I do when maxrate < bitrate?

I can set maxrate = 0, but then FFmpeg v4.2 will set maxrate to bitrate * 2 and target to 50%. FFmpeg 3.4 will set maxrate = bitrate and a target at 100% (as far as I can tell).

If I leave maxrate < bitrate, FFmpeg 3.4 will set maxrate = bitrate. FFmpeg 4.2 will give an error.

Looking at obs-ffmpeg-nvenc.c, the option "max_bitrate" do not seem to be used...

I'm tempted to set it to 0.

Reply:
0 sounds fine to me.

If the user puts something impossible in, letting ffmpeg choose limits it thinks are appropriate for your bitrate seems fine. Even if ffmpeg changes from version to version what it thinks is best.